Cache Rewrite

ychescale9 opened the pull request:
Resolves #14 

This is a rewrite of the cache module in Kotlin. The new implementation supports **sized-based** and **time-based (time-to-live and time-to-idle)** evictions.

### Cache and Builder APIs

```kotlin
interface Cache<in Key, Value> {

    /**
     * Returns the value associated with [key] in this cache, or null if there is no
     * cached value for [key].
     */
    fun get(key: Key): Value?

    /**
     * Returns the value associated with [key] in this cache if exists,
     * otherwise gets the value by invoking [loader], associates the value with [key] in the cache,
     * and returns the cached value.
     *
     * Note that if an unexpired value for the [key] is present by the time the [loader] returns
     * the new value, the existing value won't be replaced by the new value. Instead the existing
     * value will be returned.
     */
    fun get(key: Key, loader: () -> Value): Value

    /**
     * Associates [value] with [key] in this cache. If the cache previously contained a
     * value associated with [key], the old value is replaced by [value].
     */
    fun put(key: Key, value: Value)

    /**
     * Discards any cached value for key [key].
     */
    fun invalidate(key: Key)

    /**
     * Discards all entries in the cache.
     */
    fun invalidateAll()

    /**
     * Main entry point for creating a [Cache].
     */
    interface Builder {

        /**
         * Specifies that each entry should be automatically removed from the cache once a fixed duration
         * has elapsed after the entry's creation or the most recent replacement of its value.
         *
         * When [duration] is zero, the cache's max size will be set to 0
         * meaning no values will be cached.
         */
        fun expireAfterWrite(duration: Long, unit: TimeUnit): Builder

        /**
         * Specifies that each entry should be automatically removed from the cache once a fixed duration
         * has elapsed after the entry's creation, the most recent replacement of its value, or its last
         * access.
         *
         * When [duration] is zero, the cache's max size will be set to 0
         * meaning no values will be cached.
         */
        fun expireAfterAccess(duration: Long, unit: TimeUnit): Builder

        /**
         * Specifies the maximum number of entries the cache may contain.
         * Cache eviction policy is based on LRU - i.e. least recently accessed entries get evicted first.
         *
         * When [size] is 0, entries will be discarded immediately and no values will be cached.
         *
         * If not set, cache size will be unlimited.
         */
        fun maximumCacheSize(size: Long): Builder

        /**
         * Guides the allowed concurrent update operations. This is used as a hint for internal sizing,
         * actual concurrency will vary.
         *
         * If not set, default concurrency level is 16.
         */
        fun concurrencyLevel(concurrencyLevel: Int): Builder

        /**
         * Specifies [Clock] for this cache.
         *
         * This is useful for controlling time in tests
         * where a fake [Clock] implementation can be provided.
         *
         * A [SystemClock] will be used if not specified.
         */
        fun clock(clock: Clock): Builder

        /**
         * Builds a new instance of [Cache] with the specified configurations.
         */
        fun <K : Any, V : Any> build(): Cache<K, V>
    }
}
```

Note that currently only APIs required by `RealStore` are supported. Please let know if more APIs are required.

### Usage

```kotlin
// build a new cache
val cache = Cache.Builder.newBuilder()
            .expireAfterWrite(30, TimeUnit.MINUTES)
            .expireAfterAccess(24, TimeUnit.HOURS)
            .maximumCacheSize(100)
            .build<Long, String>()

// cache a value
cache.put(1, "dog")

// get cached value by key
cache.get(1)

// get cached value by key, using the provided loader to compute and cache the value if none exists
val value = cache.get(2) { "cat" }

// remove a cached value by key
cache.invalidate(1)

// remove all values from cache
cache.invalidateAll()
```

### Tests

`./gradlew cache:test`

100% Test Coverage:

<img width="646" alt="cache4-coverage" src="https://user-images.githubusercontent.com/11519072/71650776-93f13d80-2d6c-11ea-97d2-fe7899cb411a.png">

### Migration

All usages have been switched to use the new cache implementation.

I've also converted the remaining Java files to Kotlin, except `MultiParserTest.java` which is commented out.

----------

Will update `cache/README.md` in a followup.

ychescale9 commented on filesystem/src/main/java/com/dropbox/android/external/fs3/filesystem/FileSystemImpl.kt:
```diff
@@ -96,16 +95,37 @@ internal class FileSystemImpl(private val root: File) : FileSystem {
+ * Cache and return a [fallback] value when no value associated with the [key] exists in the cache,
+ * otherwise return the existing value.
+ */
+private fun <Key, Value> Cache<Key, Value>.get(key: Key, fallback: Value): Value {
```
This mimics Guava's `CacheLoader` behavior which `FileSystemImpl` was still using. Please let me know if you want this moved to the core `Cache` API.

digitalbuddha replied:
I think this behavior is slightly different. My understanding is that loader function was only called when the presence check fails. This allowed interactions such as cache.get(key, expensiveNetworkCall()) where the function call is lazy. Could we respect the behavior within the cache rewrite? Primarily I want to be able to call store.get(key, loader) N number of times concurrently and only have loader called once

ychescale9 replied:
Sure I'll add a `get(key: Key, loader: () -> Value): Value` overload.

ychescale9 replied:
I've now added cache loader support.

digitalbuddha commented:
Wow this is incredible! Thank you so much for all the hard work. I left one comment about cache loader otherwise looks good from first glance. After the holidays we will take this through a more detailed review

ychescale9 commented on cache/src/test/kotlin/com/dropbox/android/external/cache4/CacheLoaderTest.kt:
```diff
@@ -0,0 +1,314 @@
+
+    @Test
+    fun `get(key, loader) returns existing value when an entry with the associated key is absent initially but present after executing the loader`() =
+        runBlocking {
```
I wanted to use `TestCoroutineScope.runBlockingTest` here to advance the delays, but it doesn't seem to wait for the coroutines to finish when I'm launching with a different dispatcher: `launch(Dispatchers.Default)` which I need to be able to test concurrency. Is there a better way to test this?

digitalbuddha replied:
@yigit has strategies for testing async coroutines maybe he can chime in

yigit commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/CacheEntry.kt:
```diff
@@ -0,0 +1,12 @@
+ */
+internal data class CacheEntry<Key, Value>(
+    val key: Key,
+    @Volatile var value: Value,
```
as far as i can read, these are always accessed inside a lock which will put a memory barrier hence you should not need volatile for these.

ychescale9 replied:
Hmm I’m not sure I follow. These entries live in the ConcurrentHashMap and a couple of synchronized LinkedHashSet but we are updating these values directly (eg when replacing the value for an existing key) instead of recreating and putting a new entry into the map. And currently only get(key: Key, () -> Value): Value is protected by an explicit synchronized block to prevent calling the potentially expensive loader function multiple times concurrently. So I think those mutable fields in the CacheEntry should be the only things we need to protect from concurrent access and hence the volatile. Am I missing something here?

eyalgu replied:
Although using access to `ConcurrentHashMap` as a memory barrier should be enough, we do not write `Entry` back into the map after updating it so I don't believe we relay on it. If we wanted to drop `Volatile` here we would probably need to make these changes.

Whatever we end up with here (either volatile or using the map as a barrier), we should add some documentation here on why we're reusing entries (to be efficient) and document the concurrency policy if it's not obvious.

ychescale9 replied:
Using `map.put(...)` as a barrier feels a bit subtle. Should we use an explicit lock to synchronize all updates to the `CacheEntry` objects?

ychescale9 replied:
I've studied this a bit more and now I believe explicit lock or `volatile` shouldn't be necessary :)

Will remove `volatile` and add documentation about reuse.

yigit commented:
thanks for doing this btw, looks pretty good!

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/Cache.kt:
```diff
@@ -0,0 +1,37 @@
+     * the new value, the existing value won't be replaced by the new value. Instead the existing
+     * value will be returned.
+     */
+    fun get(key: Key, loader: () -> Value): Value
```
The FileSystemImpl.kt was using the CacheLoader implementation from guava so I added this as suggested by @digitalbuddha

eyalgu replied:
lets also document that loader needs to be quick (i.e it's not a suspend function).

digitalbuddha replied:
Since we want to be a replacement for Store3, I figured we should respect this api in case there are consumers of the cache only

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/CacheBuilder.kt:
```diff
@@ -0,0 +1,101 @@
+     * has elapsed after the entry's creation, the most recent replacement of its value, or its last
+     * access.
+     *
+     * When [duration] is zero, the cache's max size will be set to 0
```
This is the same behavior as Guava and yes it’s primarily for testing.

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/CacheBuilder.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    companion object {
+        internal const val UNSET_LONG: Long = -1
```
By default it’s -1 which means no size limit. See DefaultCacheTest.kt

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    override fun get(key: Key, loader: () -> Value): Value {
+        synchronized(this) {
```
Ack.

eyalgu replied:
not super granular, but I guess we can improve it in the future to be key based if we want to.

digitalbuddha replied:
Could we verify if old behavior blocked per key or not? Ideally we should be granular since an app might have a single store used among many features

ychescale9 replied:
Added key-based synchronization with new tests.

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,241 @@
+        if (existingEntry != null) {
+            // cache entry found
+            existingEntry.recordWrite(nowNanos)
+            existingEntry.value = value
```
I think you can add `cacheEntries[key] = existingEntry` to create a memory barrier using the map access.

digitalbuddha replied:
any reason not to make the loader `suspend` rather than blocking? Could be follow up

ychescale9 replied:
There's no `CoroutineScope` to execute the `suspend` function. If we made the `get(key, loader)` API itself `suspend` we will need to use `Mutex` which is also part of `kotlinx-coroutines-core`  for synchronization. Or do we want to introduce coroutines library dependency to `cache`?

eyalgu commented on cache/src/test/kotlin/com/dropbox/android/external/cache4/CacheLoaderTest.kt:
```diff
@@ -0,0 +1,314 @@
+        computedValue: Value,
+        executionTime: Long
+    ): () -> Value {
+        return spy<() -> Value>().apply {
```
Can we avoid using mocks and spys here all you need is a simple invocation counter:
```
class TestLoader<T>(private val block: () -> T) {
  var invokeCount = 0
  operation fun invoke(): T {
    invokeCount++
    return block() 
  }
```

ychescale9 commented:
@eyalgu @yigit Thanks for the review! Will address those comments soon.

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,269 @@
+        return keyBasedLocks.getLock(key).withLock {
+            action()
+        }.also {
+            keyBasedLocks.remove(key)
```
OTTOMH I can't think of a better way than a to add a counter to each lock entry in the map and have a global lock for adding/removing entries (that is aside from using Guava's ConcurrentHashMultiset which defeats the porpuse of rewriting this cache)
```
val LockMapSynchronizer = Any()
val keyBasedLocks = HashMap<Key, Pair<Lock, Int>>

private fun <Key> MutableMap<Key, Pair<Lock, Int>>.getLock(key: Key): Lock {
    synchronzied(lockMapSynchronizder) {
        val entry = get(key)
        if (entry != null) {
            entry.second++
            return entry.first
        }
        map.put(key, Pair(ReentrantLock(), AtomicInt(1)))
        return get(key).first
    }
}

private fun <Key> MutableMap<Key, Pair<Lock, Int>>.removeLock(key: Key) {
    synchronzied(lockMapSynchronizder) {
        val entry = get(key) ?: return
        entry.second--
        if (entry.second == 0) {
            remove(key)
        }
    }
}
```

eyalgu replied:
alternatively, we can take an alternative approach and not use the key as the synchronizer. we can instead use lock striping: create an array of 8/16/254/whatever entries of `Lock`. hash the key and grab the lock that matches the key's hash (mod size of array). if you lazy init this array you'll need to sync that too

eyalgu replied:
what does guava do here?

ychescale9 replied:
As far as I can tell Guava [doesn't synchronize loader execution by key](https://github.com/google/guava/blob/28b36c80447729d1edeffedd70dc4a7ac91e76ce/guava/src/com/google/common/cache/LocalCache.java#L2154-L2156), but relies on [key hash](https://github.com/google/guava/blob/28b36c80447729d1edeffedd70dc4a7ac91e76ce/guava/src/com/google/common/cache/LocalCache.java#L3950-L3951) and [Segments](https://github.com/google/guava/blob/28b36c80447729d1edeffedd70dc4a7ac91e76ce/guava/src/com/google/common/cache/LocalCache.java#L1829) to guard concurrent `get` in general but on a less granular level.

ychescale9 replied:
> you need some sort of reference counting to make this work. you would have:
> thread 1 takes lock.
> thread 2 gets lock from map and waits
> thread 1 finishes and removes key
> thread 3 comes in and does not find a lock in map. proceeds with a new lock
> thread 2 resumes on old lock

Thanks for this failing case. I'll try to implement the reference counting lock map as suggested and add more tests.

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,269 @@
+     * otherwise associates a new [Lock] with the specified [key] in the map and returns the new lock.
+     */
+    private fun <Key> MutableMap<Key, Lock>.getLock(key: Key): Lock {
+        val lock: Lock? = get(key) ?: put(key, ReentrantLock())
```
On Android `putIfAbsent` is only available on API 24+, so I implemented the same logic here by checking `get(key)` first.

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/CacheEntry.kt:
```diff
@@ -0,0 +1,18 @@
+ * Note that we assume cache entries are stored in thread-safe collections such as ConcurrentHashMap.
+ *
+ */
+internal data class CacheEntry<Key, Value>(
```
Ack

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt:
```diff
@@ -0,0 +1,68 @@
+
+    /**
+     * Executes the given [action] under a lock associated with the [key].
+     * When called concurrently, all actions are executed sequentially for the same [key].
```
sequentially may imply FIFO or fairness that we do not intend to guarantee. lets stick with a more neutral language

```suggestion
     * When called concurrently, all actions associated with the same [key] are mutually exclusive.
```

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt:
```diff
@@ -0,0 +1,68 @@
+
+private class LockEntry(
+    val lock: Lock,
+    val counter: AtomicInteger
```
all access to the counter are under lock. no need to use atomic int
```suggestion
    val counter: Int
```

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt:
```diff
@@ -0,0 +1,68 @@
+            }
+            // create a new lock entry with counter set to 1 indicating the lock is used by 1 thread
+            keyBasedLocks[key] = LockEntry(ReentrantLock(), AtomicInteger(1))
+            return keyBasedLocks[key]!!.lock
```
```
val lockEntry = keyBasedLocks[key] ?: LockEntry(ReentrantLock(), 0)
lockEntry.counter++
return lockEntry.lock
```

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt:
```diff
@@ -0,0 +1,68 @@
+            // decrement the counter to indicate the lock is no longer needed for this thread,
+            // then remove the lock entry from map if no other thread is still holding this lock
+            val lockEntry = keyBasedLocks[key] ?: return
+            if (lockEntry.counter.decrementAndGet() == 0) {
```
```suggestion
            lockEntry.counter--
            if (lockEntry.counter == 0) {
```

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,262 @@
+    /**
+     * A key-based synchronizer for running cache loader.
+     */
+    private val loaderSynchronizer = KeyedSynchronizer<Key>()
```
```suggestion
    private val loadersSynchronizer = KeyedSynchronizer<Key>()
```

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt:
```diff
@@ -0,0 +1,68 @@
+     * Executes the given [action] under a lock associated with the [key].
+     * When called concurrently, all actions are executed sequentially for the same [key].
+     */
+    fun <T> run(key: Key, action: () -> T): T {
```
```suggestion
    fun <T> synchronizedFor(key: Key, action: () -> T): T {
```

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,262 @@
+            // cache entry found
+            recordWrite(existingEntry, nowNanos)
+            existingEntry.value = value
+            cacheEntries[key] = existingEntry
```
can you add a comment about memory barrier to explain for future readers why this line is necessary

ychescale9 commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/KeyedSynchronizer.kt:
```diff
@@ -0,0 +1,67 @@
+            // save the lock entry to the map if it has just been created
+            if (keyBasedLocks[key] == null) {
+                keyBasedLocks[key] = lockEntry
+            }
```
`putIfAbsent` is only available on API 24+.

digitalbuddha commented:
Hello sir, could you run some benchmarks for us to have an idea of whether we should invest in optimizing after landing? There is a library caffeine that has a few benchmarks it used to compare itself to guava, could you add your impl and compare as well?
https://github.com/ben-manes/caffeine/wiki/Benchmarks

https://github.com/ben-manes/caffeine/tree/master/caffeine/src/jmh/java/com/github/benmanes/caffeine/cache

Thank you kindly

ychescale9 commented:
@digitalbuddha sure I’ll run some benchmarks locally and share results here tomorrow. I don’t expect our results to be great though as currently it’s just a wrapper around ConcurrentHashMap so perf will be no match for Caffeine or Guava 😅, although there should still be some low hanging fruits with current implementation as we didn’t really focus on perf in this pass.

Also would this cache module be positioned as a generic solution or something primarily targeted at mobile? When would a user use this cache directly without store?

digitalbuddha commented:
"I don’t expect our results to be great though as currently it’s just a wrapper around ConcurrentHashMap so perf will be no match for Caffeine or Guava 😅, although there should still be some low hanging fruits with current implementation as we didn’t really focus on perf in this pass."
I don't think we are concerned with speed too much, more want to know where we stand to properly scope future work.


Also would this cache module be positioned as a generic solution or something primarily targeted at mobile? 
Mobile only as far as I can tell from users, I can see it being used for kotlin web as well.


 When would a user use this cache directly without store?
The filesystem module is used by NYT and a few personal projects of mine. I have also used the cache independently whenever I did not want to pull all of guava in. I don't think we need to be the fastest but more make sure that we don't have any glaring issues.  


Thank you for all the incredible work in this diff

ychescale9 commented:
I've done some benchmarks locally by adding store/cache to caffeine's benchmark setup.

### Machine specs
MacBook Pro
Processor: Intel Core i7 2.2GHz (4 cores)
Memory: 16Gb DDR3
OS: macOS Catalina

### Compute Benchmark

This benchmark focuses on the `CacheLoader` API. Our implementation is significantly slower than **ConcurrentHashMap**, **Guava** and **Caffeine** possibly due to the much more granular per-key locking we are doing (and possibly the way we implement lock maps synchronization) but I haven't verified the impact of locks here. Perhaps we could look at implementing lock striping in the future as @eyalgu suggested.

![ComputeBenchmark](https://user-images.githubusercontent.com/11519072/71759037-5c64cb80-2efb-11ea-9976-19dce7247284.png)

### Get Put Benchmark

These benchmarks focus on read / write performance with `get(key)` and `put(key, value)`.

We are better than **ElasticSearch** and **ExpiringMap_Lru** but significantly slower than the others (most of which implement custom segment-based data structures).

![GetPutBenchmark_ReadOnly](https://user-images.githubusercontent.com/11519072/71759120-528f9800-2efc-11ea-9bc7-426282678370.png)
![GetPutBenchmark_ReadWrite](https://user-images.githubusercontent.com/11519072/71759121-56bbb580-2efc-11ea-8907-a3aa2ee002de.png)
![GetPutBenchmark_WriteOnly](https://user-images.githubusercontent.com/11519072/71759124-591e0f80-2efc-11ea-93c4-bd21ba943f04.png)

### Eviction Benchmark

This benchmark evaluates the eviction performance of a cache. The cache is pre-populated for a 100% eviction rate to mimic worst case behavior.

Surprisingly we're very close to **Caffeine** here.

![EvictionBenchmark](https://user-images.githubusercontent.com/11519072/71759153-d77ab180-2efc-11ea-927e-efe11d25605a.png)

----------

Once this is available in the snapshot build I can publish these as a Caffeine fork that focuses on **Store Cache** if people are interested. Or we could even set up **JMH** or **Jetpack Benchmark** properly in our repo.

ychescale9 commented:
I've made a couple of changes while playing with benchmarks:

- Only process the `AccessQueue` if `expiresAfterAccess` is enabled.
- Added support for configuring the `concurrencyLevel`, as the some of the benchmarks require setting a concurrency level of 64 and we were hardcoding it to 4.

I've also made some other misc improvements / cleanups but since they are unrelated to this PR I'll send them once this is merged.

ychescale9 commented:
Some of the concurrency tests with real delays can be flaky at times. Perhaps we should allow some automatic retries for these tests? Again please let me know if there are better ways to write these tests.

digitalbuddha commented:
Waiting for 1 more approver before we merge

ben-manes commented:
Hello,

It appears that you have a memory leak in your implementation due to non-atomic writes. It is important to remember that threads may be preempted at any time and may be delayed significantly from running again by the OS's scheduler. This means that interleaving of atomic operations will occur with enough concurrency.

The simplest case is a `put` racing with `invalidateAll`:
Thread1: Inserts (K, V), adds to the queues (`recordWrite`), is preempted before writing to the map. 
Thread2: Clears the queues and map
Thread1: Inserts into the map
Problem: The entry is no longer eligible for removal by eviction unless it is relinked into the queues as a side-effect of being read or updated. If a "one-hit wonder", it will never be accessed again and over time cause `evictEntries` to NPE when the `accessQueue` becomes empty but the cache still exceeds its capacity threshold. 

Similar problems occur with writes interleaving eviction when an entry is added, `recordWrite` completes, another thread's `evictEntries` discards it, and the entry is added to the map.

The simplest solution would be to make the locking coarser to cover all operations on the map and queues by a single lock. The per-key locking can remain above that.

ychescale9 commented:
Thanks @ben-manes!

> The simplest solution would be to make the locking coarser to cover all operations on the map and queues by a single lock. The per-key locking can remain above that.

Does this mean that all the `get(key)` and `put(key, value)` calls will be effectively sequential (mutually exclusive)? Performance isn't our focus here, but is there any way to eliminate the memory leak without adding too much synchronization cost?

Would it be enough if we tried to address any inconsistency between the `accessQueue` and the map (i.e. remove any entries that are in the map but **not**  in the `accessQueue`) before performing evictions?

```kotlin
private val evictionLock: Lock = ReentrantLock()

private fun evictEntries() {
    if (!evictsBySize) {
        return
    }

    checkNotNull(accessQueue)

    evictionLock.withLock {
        // clean up leaked cache entries before performing evictions
        if (accessQueue.size < cacheEntries.size) {
            val iterator = cacheEntries.iterator()
            for (item in iterator) {
                val cacheEntry: CacheEntry<Key, Value> = item.value
                if (!accessQueue.contains(cacheEntry)) {
                    iterator.remove()
                    writeQueue?.remove(cacheEntry)
                }
            }
        }

        while (cacheEntries.size > maxSize) {
            val entryToEvict = accessQueue.first()
            cacheEntries.remove(entryToEvict.key)
            writeQueue?.remove(entryToEvict)
            accessQueue.remove(entryToEvict)
        }
    }
}
```

Really appreciate your help! Wasn't really expecting Caffeine's author to be reviewing my naive attempt at writing a cache in Kotlin 😅

ben-manes commented:
> Does this mean that all the get(key) and put(key, value) calls will be effectively sequential (mutually exclusive)?

Yes. Currently the synchronization on `accessQueue` for every read means it already has a sequential bottleneck. Since a synchronized LRU cache is faster (via `LinkedHashMap`) and you are targeting Android, you probably don't need anything fancier.

> Performance isn't our focus here, but is there any way to eliminate the memory leak without adding too much synchronization cost?

In a synchronized lru, the trick used by memcached is to skip updating the LRU if the item was touched within the last 60 seconds. There are of course other designs, but that is probably the easiest optimization for you to apply.

> Would it be enough if we tried to address any inconsistency between the accessQueue and the map (i.e. remove any entries that are in the map but not in the accessQueue) before performing evictions?

Yes, something like that would work. To be safe you'd want to remove entries that are disjoint (either not in the queue or not in the map), for both queues. Assuming this happens rarely, it is probably good enough as these will be small caches.

ychescale9 commented:
Thanks Ben!

pavlospt commented on cache/src/test/kotlin/com/dropbox/android/external/cache4/testutil/ConcurrencyTestRule.kt:
```diff
@@ -0,0 +1,47 @@
+        val annotation: ConcurrencyTest =
+            description.getAnnotation(ConcurrencyTest::class.java) ?: return base
+        val times: Int = annotation.retries
+        require(times > 0) { "Number of retires must be greater than 1." }
```
@ychescale9 possible typo here: `retires`

ychescale9 replied:
Thanks!

pavlospt commented on cache/src/test/kotlin/com/dropbox/android/external/cache4/testutil/ConcurrencyTestRule.kt:
```diff
@@ -9,26 +9,26 @@ import org.junit.runners.model.Statement
             description.getAnnotation(ConcurrencyTest::class.java) ?: return base
-        val times: Int = annotation.retries
-        require(times > 0) { "Number of retires must be greater than 1." }
+        val maxRetires: Int = annotation.retries
```
@ychescale9 there is the same typo on the variable name here as well :) `retires`

ychescale9 replied:
:man_facepalming: Thanks! And no I'm not considering retirement just yet...

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,328 @@
+                        }
+                    }
+                }
+            } else if (queueSize > cacheEntrySize) {
```
by having both these checks it implies that you can also have a condition where you have the same number of cache entries in the map and the queue but they do not match.

Assuming that we always record the entry in the queues before adding it to the map (which is the case) we will also actually introduce a bug here where in the following interleaving:
thread 1 records 2 entry in the queue
thread 2 calls cleanup and removes that entry from the queue because it doesn't have a counterpart in the map
thread 1 adds the entry into the map
the next call to cleanup will evict the new entry from the map

Luckily if we always first update the queues then update the map we only need to remove entries in the map that do not have matching entries in the queue (as the only way that they were removed is if the entry was evicted from the cache but was then re-added by the interrupted put thread).

To fix this bug I think we can just remove the `queueSize > cacheEntrySize` branch and make sure to always update the map after the queue(s)
Also, if you are in the process of adding a new entry than it may g

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,328 @@
+                // if entry's access time has been updated, perform an update to the map to create
+                // a memory barrier for synchronizing subsequent access to the map from other threads.
+                if (expiresAfterAccess) {
+                    cacheEntries[key] = it
```
in retrospect, it's probably better to not re-add entries to the map but make the `var`s volatile. the (rare) race condition we have not is:
thread 1 reads value1 and calls recordRead
thread 2 evicts value1 from cache
thread 2 adds value2 to the map
thread 1 resumes and re-adds value1 to the map, replacing the newer value2.

let's avoid this by removing all the `cacheEntry[key] = it` calls in the map and marks the `var`s in cache entry as `@Volatile`

eyalgu commented on cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt:
```diff
@@ -0,0 +1,305 @@
+                        iterator.remove()
+                        // also remove the entry from the other queue
+                        if (queue == writeQueue) {
+                            accessQueue?.remove(entry)
```
There's a hidden deadlock here. Sync set uses the set as the lock for remove. If 2 threads are evicting entries they can deadlock on each other. Remove this optimization. You will remove the entry when iterating on the next queue

eyalgu replied:
I mean remove the entire if-else removes

ychescale9 replied:
Good catch. Thanks.

ychescale9 commented:
> Sorry I found one more bug. Caching is hard

I thought I knew what I signed up for when I decided to do this...

eyalgu commented:
@yigit I unblocked only needs your ok now

digitalbuddha commented:
merging tonight unless any qualms

digitalbuddha commented:
Could you resolve conflicts and then I'll merge. Thank you

ychescale9 commented:
Resolved the conflicts.

As a result of making the `Cache`'s `Key` and `Value` bounded by `Any` to make sure neither can be null, these changes propagated to the `Store` and `StoreBuilder` APIs. I just found out that these have not been updated in `DiskReader`, `DiskWriter` and `Persister` (maybe more). Maybe we can clean up these in a followup (after alpha01 release)?

- I've also removed the old `cache/README.md` and I'll write a new one later.
- Also removed the last remaining java file `MultiParserTest.java`

digitalbuddha commented:
Great job!! We will be doing our first alpha release today along with a blog post 😁
